tests/components/statistics/test_sensor.py

```diff
@@ -1701,3 +1702,60 @@ async def test_device_id(
     statistics_entity = entity_registry.async_get("sensor.statistics")
     assert statistics_entity is not None
     assert statistics_entity.device_id == source_entity.device_id
+
+
+async def test_update_before_load(recorder_mock: Recorder, hass: HomeAssistant) -> None:
+    """Verify that updates happening before reloading from the database are handled correctly."""
+
+    current_time = dt_util.utcnow()
+
+    # enable and pre-fill the recorder
+    await hass.async_block_till_done()
+    await async_wait_recording_done(hass)
+
+    # enable and pre-fill the recorder
+    await hass.async_block_till_done()
+    await async_wait_recording_done(hass)
+
+    with (
+        freeze_time(current_time) as freezer,
+    ):
+        for value in VALUES_NUMERIC_LINEAR:
+            hass.states.async_set(
+                "sensor.test_monitored",
+                str(value),
+                {ATTR_UNIT_OF_MEASUREMENT: UnitOfTemperature.CELSIUS},
+            )
+            await hass.async_block_till_done()
+            current_time += timedelta(seconds=1)
+            freezer.move_to(current_time)
+
+        await async_wait_recording_done(hass)
+        # create the statistics component, get filled from database
+        assert await async_setup_component(
+            hass,
+            "sensor",
+            {
+                "sensor": [
+                    {
+                        "platform": "statistics",
+                        "name": "test",
+                        "entity_id": "sensor.test_monitored",
+                        "state_characteristic": "average_step",
+                        "max_age": {"seconds": 10},
+                    },
+                ]
+            },
+        )
+        # this value is going to be ignored, since loading from the database hasn't finished yet
```

Comment:
Am I missing something or is this not the opposite of what should happen?
It should grab all the relevant states from the recorder and then start collecting state changes from events?

Reply:
Currently, we have a race condition. New values are added to the queue before the values loaded from the database were loaded. As @emontnemery proposed we are now accepting new values only after loading data from the database succeeded. This means that whatever comes in before will be ignored.

Reply:
Any risk this also causes a race in the test e.g. the setup goes faster than we can write a new state (that should be ignored)?

Reply:
@gjohansson-ST: This is something I can't answer for sure, since it would require a good understanding of the framework and what the involved functions `async_setup_component` and `async_set` are doing. I definitely lack that knowledge, and I don't want to dig through all that code to get it.

Having said that, I have executed the tests many times without failure (which of course is not even remotely a proof that there are no race conditions). 

Furthermore, we can safely assume that loading something from the database will involve IO and hence a handover to another async activity will happen. Assuming that the `async_set` function will become active, it will at least not be interrupted within the code within the statistics module, since there is no I/O, sleep or whatever that would make it hand over control. But if the `async_set` function would also do something that would hand over control again (either now or after future changes), or there are other functions executed first and the database activity is given control again, we could probably see cases where the database load operation would be finished first.

I don't know how I could guarantee the right order of events. We would have to block the database activity until we are sure that the set activity is done.

I will adapt the test to allow both possible outcomes (value was ignored, value was added, but at the end). This can only be tested with the time-based averages, since the other functions don't change depending on the order of the buffer.